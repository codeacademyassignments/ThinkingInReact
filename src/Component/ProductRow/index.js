import React from 'react';

const ProductRow = (props)=>{
const name = props.inStock ? props.name : <span id='name' style={{color:'red'}}>{props.name}</span>;
return (<tr>
  <td >{name}</td>
  <td> {props.price}</td>
  </tr>);
}

export default ProductRow;