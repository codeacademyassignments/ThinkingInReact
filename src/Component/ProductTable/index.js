import React, {Component} from 'react';
import ProductCategoryRow from '../ProductCategoryRow';
import ProductRow from '../ProductRow';


class ProductTable extends Component{
  sortProductsOnCategory = (products)=>{
    return products.slice().sort(function(a,b){
      return a.category.toString().localeCompare(b.category.toString());
    })
  }

  generateProductTableRows = (productArray,searchKey)=>
  {
    const rows = [];
    let curr = null;
    productArray.forEach((product,index)=>{
      if(product.name.indexOf(searchKey)===-1)return;
      if(!product.stocked && this.props.inStockOnly)return;
      if(curr!==product.category)
      {
        rows.push(<ProductCategoryRow key = {index+product.category} category={product.category}/>);
        curr = product.category;
      }
      rows.push(<ProductRow key = {index+product.name} name = {product.name}  price={product.price} inStock={product.stocked}/>)
    })
    return rows;
  }
  render(){
    const products = this.props.products;
    const key = this.props.searchKey;
    const sortedProductArray = this.sortProductsOnCategory(products);

    const tableRows = this.generateProductTableRows(sortedProductArray,key);
    return (
      <table>
        <thead>
          <tr>
            <th>Name</th>
            <th>Price</th>
          </tr>
        </thead>
        <tbody>
        {tableRows}
        </tbody>

      </table>)
  }
}

export default ProductTable;