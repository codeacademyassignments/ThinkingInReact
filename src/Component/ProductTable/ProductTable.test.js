import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import ProductTable from '.';
import ProductCategoryRow from '../ProductCategoryRow';
import ProductRow from '../ProductRow';

const sortedProducts = [
    
  {category: "Electronics", price: "$99.99", stocked: true, name: "iPod Touch"},
  {category: "Electronics", price: "$399.99", stocked: false, name: "iPhone 5"},
  {category: "Electronics", price: "$199.99", stocked: true, name: "Nexus 7"},
  {category: "Sporting Goods", price: "$29.99", stocked: false, name: "Basketball"},
  {category: "Sporting Goods", price: "$49.99", stocked: true, name: "Football"},
  {category: "Sporting Goods", price: "$9.99", stocked: true, name: "Baseball"} 
];
const unsortedProducts = [
  {category: "Sporting Goods", price: "$29.99", stocked: false, name: "Basketball"},
  {category: "Electronics", price: "$99.99", stocked: true, name: "iPod Touch"},
  {category: "Electronics", price: "$399.99", stocked: false, name: "iPhone 5"},
  {category: "Sporting Goods", price: "$49.99", stocked: true, name: "Football"},
  {category: "Electronics", price: "$199.99", stocked: true, name: "Nexus 7"},
  {category: "Sporting Goods", price: "$9.99", stocked: true, name: "Baseball"} 
];
describe('snapshot testing',()=>{
  it('snapshot should match to previous snapshot',()=>{
    const tree = renderer.create(<ProductTable searchKey = 'iPod' products={sortedProducts} inStockOnly={false}/> );
    expect(tree).toMatchSnapshot();
  });

})

describe('sortProductOnCategory',()=>{
  const wrapper = shallow(<ProductTable searchKey = '5' products={unsortedProducts} inStockOnly={false}/> );
  it('should sort the product array passed',()=>{
    //expect({abc:'jksjk'}).toEqual({abc:'jksjk'});
    expect(wrapper.instance().sortProductsOnCategory(unsortedProducts)).not.toEqual(unsortedProducts);
    expect(wrapper.instance().sortProductsOnCategory(unsortedProducts)).toEqual(sortedProducts);
  });

})

describe('generateProductTableRow',()=>{
  it('should generate appropriate jsx',()=>{
  const productTableRows = [
  <ProductCategoryRow category="Sporting Goods" key={'4Sporting Goods'}/>,
  <ProductRow inStock={true} name="Football" price="$49.99" key={'4Football'}/>];
  const wrapper = shallow(<ProductTable searchKey = 'Foot' products={unsortedProducts} inStockOnly={false}/>);
  expect(wrapper.instance().generateProductTableRows(sortedProducts,'Foot')).toEqual(productTableRows);
  });
  it('should generate appropriate jsx: empty object when keyword doesnot match with any product name',()=>{
    const wrapper = shallow(<ProductTable searchKey = '$' products={unsortedProducts} inStockOnly={false}/>);
    expect(wrapper.instance().generateProductTableRows(sortedProducts,'$')).toEqual([]);
    });
    it('should generate appropriate jsx: empty object when instockonly true and out of stock product only matches',()=>{
      const wrapper = shallow(<ProductTable searchKey = 'iPhone' products={unsortedProducts} inStockOnly={true}/>);
      expect(wrapper.instance().generateProductTableRows(sortedProducts,'iPhone')).toEqual([]);
      });
})