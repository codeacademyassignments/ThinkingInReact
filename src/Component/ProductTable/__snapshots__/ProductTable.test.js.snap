// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing snapshot should match to previous snapshot 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Price
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th
        colSpan="2"
      >
        Electronics
      </th>
    </tr>
    <tr>
      <td>
        iPod Touch
      </td>
      <td>
         
        $99.99
      </td>
    </tr>
  </tbody>
</table>
`;
